cfme/web_ui/__init__.py

```diff
@@ -1687,134 +1687,106 @@ def _select_chkboxtree_seq(cbtree, values):
             cbtree.uncheck_node(*path)
 
 
-class InfoBlock(Pretty):
-    """ A helper class for information blocks on pages
+class InfoBlock(object):
+    DETAIL = "detail"
+    FORM = "form"
+    _TITLE_CACHE = {}
 
-    This class is able to work with both ``detail`` type information blocks and
-    ``form`` style information blocks. It is invoked with a single argument describing
-    the type of blocks to be addressed and adjusts the html elements accordingly.
-
-    Args:
-        itype: The type of information blocks to address, either ``detail`` or ``form``.
-
-    Returns: Either a string if the element contains just text, or the element.
-
-    Raises:
-        exceptions.BlockTypeUnknown: If the Block type requested by itype is unknown.
-        exceptions.ElementOrBlockNotFound: If the Block or Key requested is not found.
-        exceptions.NoElementsInsideValue: If the Key contains no elements.
-
-    An InfoBlock only needs to know the **type** of InfoBlocks you are trying to address.
-    You can then return either text, the first element inside the value or all elements.
+    def __init__(self, title):
+        self.title = title
+        self._type = None
+        self._member_cache = {}
 
-    Usage:
+    @classmethod
+    def create(cls, title, detail):
+        if title not in cls._TITLE_CACHE:
+            cls._TITLE_CACHE[title] = InfoBlock(title)
+        return InfoBlock(title).member(detail)
 
-        block = web_ui.InfoBlock("form")
+    @property
+    def type(self):
+        if self._type is None:
+            self.root  # To retrieve it
+        return self._type
 
-        block.text('Basic Information', 'Hostname')
-        block.element('Basic Information', 'Company Name')
-        block.elements('NTP Servers', 'Servers')
+    @property
+    def root(self):
+        detail_loc = version.pick({
+            '5.3': '//table//th[contains(., "{}")]/../../../..'.format(self.title),
+            version.LOWEST:
+            '//div[@class="modbox"]/h2[@class="modtitle"][contains(., "{}")]/..'.format(self.title)
+        })
+        form_loc = version.pick({
+            version.LOWEST: '//fieldset/p[@class="legend"][contains(., "{}")]/..'.format(self.title)
+        })
+        try:
```

Comment:
Would it not be better to do a generic element search for ```//div[@class="modbox"]/h2[@class="modtitle"]``` or ```//fieldset/p[@class="legend"]``` to determine the block type first, otherwise you could receive a BlockTypeUnknown for a block which is valid but the user has the title wrong.